bottlecap/src/telemetry/listener.rs

```diff
@@ -30,22 +46,27 @@ impl HttpRequestParser {
     /// It will also error if the headers cannot be parsed.
     ///
     /// Or if the body cannot be parsed.
-    pub fn from_stream(mut stream: &TcpStream) -> Result<HttpRequestParser, Box<dyn Error>> {
-        stream.set_nonblocking(true)?;
+    pub async fn from_stream(stream: &mut TcpStream) -> Result<HttpRequestParser, TcpError> {
         let mut parser = HttpRequestParser {
             headers: HashMap::new(),
             body: String::new(),
         };
 
         let mut headers_buf = [0u8; HEADERS_BUFFER_SIZE];
         loop {
-            match stream.read(&mut headers_buf) {
+            // select here
+            match stream.read(&mut headers_buf).await {
+                Ok(0) => {
+                    error!("astuyve Connection closed by client");
+                    return Err(TcpError::Close("Connection closed by client".to_string()));
+                }
```

Comment:
```suggestion
            match stream.read(&mut headers_buf).await {
```
Do we really need to check a connection close?

Reply:
Yes we do, this fixes a bug in our current implementation where we'll sit around forever holding this socket open for this client until we exit (even after the client hung up).

0 is success/`OK` result so it falls into our blocking `wait` state. But if you read the docs for `stream.read` it says `0 means EOF`, so the client closed the connection.

Reply:
ahh you're right, we need to match on this.

Reply:
Fixed